### PR TITLE
i#3544: RV64: Mangle pc-relative addressing destining to tp

### PR DESCRIPTION
**Pull request by ziyao233**
Reuse mangle_stolen_reg_and_tp_reg() to simplify register spilling.

This handles code sequence like

```
_start:
        lla             tp,             slot
	lw              a0,             0(tp)
	li              a7,             93
	ecall
slot:   .long           0x5a
```

Issue: #3544

### CODE REVIEW

File: core/arch/riscv64/mangle.c [hunk 1]
```diff
@@ -619,22 +619,29 @@ mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
...
+    if (uses_reg_stolen) {
         PRE(ilist, next_instr,
             instr_create_save_to_tls(dcontext, DR_REG_A0, TLS_REG_STOLEN_SLOT));
         PRE(ilist, next_instr,
             instr_create_restore_from_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
+    } else if (uses_reg_tp) {
+        PRE(ilist, next_instr,
+            instr_create_save_to_tls(
+                dcontext, DR_REG_A0,
+                os_tls_offset(os_get_app_tls_base_offset(TLS_REG_LIB))));
```
Comment (ksco) [inline core/arch/riscv64/mangle.c:642]: `os_tls_offset()` should be unnecessary for RISC-V-specific code, so I think it's a bit cleaner not to use it. (There are some uses of this function in #7062 too, please change that too if it makes sense.)

**Review by ksco — APPROVED** [review]

File: core/arch/riscv64/mangle.c (hunk 1 above)
Comment (ziyao233) [reply core/arch/riscv64/mangle.c:642]: > `os_tls_offset()` should be unnecessary for RISC-V-specific code, so I think it's a bit cleaner not to use it. (There are some uses of this function in #7062 too, please change that too if it makes sense.)

Though `os_tls_offset()` is just nop on RISC-V, I think keeping its usage sync with other architectures helps understand how the code works. Dropping it does save a few bytes but brings assumption of another module, which seems not a good practice.

File: core/arch/riscv64/mangle.c [hunk 2]
```diff
@@ -619,22 +619,29 @@ mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
...
 
     insert_mov_immed_ptrsz(dcontext, (ptr_int_t)tgt, dst, ilist, next_instr, NULL, NULL);
 
-    if (instr_uses_reg(instr, dr_reg_stolen)) {
+    if (uses_reg_stolen) {
         PRE(ilist, next_instr,
             instr_create_save_to_tls(dcontext, DR_REG_A0, TLS_REG_STOLEN_SLOT));
         PRE(ilist, next_instr,
             instr_create_restore_from_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
+    } else if (uses_reg_tp) {
```
Comment (derekbruening) [inline core/arch/riscv64/mangle.c:638]: Should we try to share the general mangle_cbr_stolen_reg_and_tp_reg() code to avoid these errors of overlooking one case?

File: core/arch/riscv64/mangle.c [hunk 3]
```diff
@@ -619,22 +619,29 @@ mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
...
+        PRE(ilist, next_instr,
+            instr_create_save_to_tls(
+                dcontext, DR_REG_A0,
+                os_tls_offset(os_get_app_tls_base_offset(TLS_REG_LIB))));
+        PRE(ilist, next_instr,
+            instr_create_restore_from_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
```
Comment (derekbruening) [inline core/arch/riscv64/mangle.c:644]: Is there a test we can add these cases onto, to exercise these code paths?

**Review by derekbruening — APPROVED** [review]

File: core/arch/riscv64/mangle.c (hunk 1 above)
Comment (ksco) [reply core/arch/riscv64/mangle.c:642]: > I think keeping its usage sync with other architectures helps understand how the code works.

Well, it’s not always used in other arch-specific code too, so there is no synchronization to keep, also os_get_app_tls_base_offset() was already used several times in RISC-V-specific code without os_tls_offset().

> Dropping it does save a few bytes but brings assumption of another module, which seems not a good practice.

Well, it’s arch-specific code, it is where you make all kinds of assumptions.

File: core/arch/riscv64/mangle.c (hunk 2 above)
Comment (ziyao233) [reply core/arch/riscv64/mangle.c]: Thanks for the hint! I think reuse `mangle_special_registers()` is better.

File: core/arch/riscv64/mangle.c (hunk 1 above)
Comment (ziyao233) [reply core/arch/riscv64/mangle.c]: It makes sense to me.

In the new commit, `mangle_special_registers()` is reused for context saving, I think this could be marked as resolved.

File: core/arch/riscv64/mangle.c (hunk 3 above)
Comment (ziyao233) [reply core/arch/riscv64/mangle.c]: I did a quick dig through and found there is not existing test for code mangling on riscv64, so I prefer to add a test (I guess it should be `suite/tests/common/allasm_riscv64_mangle`?) for all corner cases in `riscv64/mangle.c` in a separate PR.

For now, the code sequence mentioned in the description is a simple test for this.

File: core/arch/riscv64/mangle.c [hunk 4]
```diff
@@ -612,37 +612,19 @@ instr_t *
...
-    ASSERT(opnd_is_rel_addr(instr_get_src(instr, 0)));
-    bool uses_reg_stolen = instr_uses_reg(instr, dr_reg_stolen),
-         uses_reg_tp = instr_uses_reg(instr, DR_REG_TP);
 
-    if (uses_reg_stolen || uses_reg_tp) {
-        dst = opnd_create_reg(DR_REG_A0);
-        PRE(ilist, next_instr,
-            instr_create_save_to_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
-    }
+    mangle_special_registers(dcontext, ilist, instr, next_instr);
```
Comment (derekbruening) [inline core/arch/riscv64/mangle.c:620]: Best to get the return value and pass it on in case something inside here removes next_instr.

File: core/arch/riscv64/mangle.c [hunk 5]
```diff
@@ -612,37 +612,19 @@ instr_t *
...
-            instr_create_restore_from_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
-    } else if (uses_reg_tp) {
-        PRE(ilist, next_instr,
-            instr_create_save_to_tls(
-                dcontext, DR_REG_A0,
-                os_tls_offset(os_get_app_tls_base_offset(TLS_REG_LIB))));
-        PRE(ilist, next_instr,
-            instr_create_restore_from_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
-    }
+    insert_mov_immed_ptrsz(dcontext, (ptr_int_t)tgt, dst, ilist, instr_get_next(instr),
```
Comment (derekbruening) [inline core/arch/riscv64/mangle.c:626]: Why instr_get_next(instr)?  Just `instr` seems more straightforward?

**Review by derekbruening — APPROVED** [review]

File: core/arch/riscv64/mangle.c (hunk 5 above)
Comment (ziyao233) [reply core/arch/riscv64/mangle.c:626]: Yes, `instr` is removed later and inserting the mov whether before or later doesn't makes a difference.

File: core/arch/riscv64/mangle.c (hunk 4 above)
Comment (ziyao233) [reply core/arch/riscv64/mangle.c:620]: next_instr isn't used anymore in the function, so I think it's unneeded. And we cannot return a non-NULL pointer in `mangle_rel_addr()`, since we delete the original `instr` and must inform the caller with a NULL return value[1]. Maybe we should add a note of not removing `next_instr` to `mangle_special_registers()`?

[1]: https://github.com/DynamoRIO/dynamorio/blob/0ab4d54744519e01854784612a83ddc57a16b0ad/core/arch/mangle_shared.c#L1953-L1965

Comment (derekbruening) [reply core/arch/riscv64/mangle.c:620]: This seems dangerous vs future changes: IMHO either mangle_special_registers() should not return a new next_instr then, so it cannot remove it, or else this function needs to return next_instr so it can pass on the return value from mangle_special_registers().

Comment (ziyao233) [reply core/arch/riscv64/mangle.c]: I switch to use mangle_stolen_reg_and_tp_reg(), which doesn't return a new next_instr

**Comment by ziyao233** [comment]
I think this PR is ready for merging :)
